maze.pl~

```diff
@@ -0,0 +1,259 @@
+test(connected_backward) :- connected(cell(2,1), cell(2,0)), !.
+
+% Find all cells connected to/from 0,0
+test(connected_from_0_0, all(X == [cell(0,1)])) :- connected(cell(0,0), X).
+test(connected_to_0_0, all(X == [cell(0,1)])) :- connected(X, cell(0,0)).
+
+% Find all cells connected to/from 1,3
+test(connected_from_1_2, set(X == [cell(0,2), cell(1,1), cell(2,2)])) :-
+  connected(cell(1,2), X).
+test(connected_to_1_2, set(X == [cell(0,2), cell(1,1), cell(2,2)])) :-
+  connected(X, cell(1,2)).
+
+:- end_tests(connected).
+
+
+% Step 2: path_to/3
+%
+% Now create path_to/3 which takes three arguments:
+%  1. A start cell
+%  2. An end cell
+%  3. A path from the start cell to the end cell as a list.
+%
+%  Tips:
+%    1. You will probably need to use an accumulator
+%    2. It may be easier to construct the path in reverse and then reverse it
+%    3. You can use negation to eliminate paths that backtrack on themselves
+%
+% Ensure that the tests pass with no warnings.
+%
+
+% (TODO): find what is wrong with path to self.
+%path_to(From,To,Path) :- connected(From,To); connected(To,From);
+%                         connected(To,Path);
+%			 connected(From,Path).
+
+
+%path_to(X,Y,[X,Y]):-corridor(X,Y),!.
+%path_to(X,Y,[X|R]):-corridor(X,Z),path_to(Z,Y,R),!.
+%
+
+path_to(From,To,Path) :-
+	travel(From,To,[From],Q),
+	rev(Q,Path),!.
+
+travel(From,To,Path,[To|Path]) :-
+	connected(From,To),!.
+
+
+travel(From,To,Visited,Path) :-
+	connected(From,C),
+	C\==To,
+	not(mem(C,Visited)),
+	travel(C,To,[C|Visited],Path),!.
+
+
+:- begin_tests(path_to).
+
+% verify that a path from a cell to itself consists of just the one cell
+test(path_to_self) :-
+  path_to(cell(0,0), cell(0,0), [cell(0,0)]).
+
+% verify path going from 0,0 to 0,1
+test(path_from_0_0_to_0_1) :-
+  path_to(cell(0,0), cell(0,1), [cell(0,0), cell(0,1)]).
+
+% verify path going from 0,1 to 0,0
+test(path_from_0_1_to_0_0) :-
+  path_to(cell(0,1), cell(0,0), [cell(0,1), cell(0,0)]).
+
+% verify path going from 0,0 to 1,1
+test(path_from_0_0_to_1_1) :-
+  path_to(cell(0,0), cell(1,1), [cell(0,0), cell(0,1), cell(1,1)]).
+
+% verify path going from 0,0 to 1,0
+test(path_from_0_0_to_1_0) :-
+  path_to(cell(0,0), cell(1,0),
+  [cell(0,0), cell(0,1), cell(1,1), cell(1,2), cell(2,2), cell(3,2), cell(3,1),
+  cell(2,1), cell(2,0), cell(1,0)]).
+
+:- end_tests(path_to).
+
+
+% Step 3: the solution
+%
+% Finally, you can put all of your work together by creating a rule
+% maze_solution/4 that takes the following arguments:
+%
+%   1. the starting position
+%   2. the ending position
+%   3. the path from the start to the end
+%   4. the length of the path
+%
+% Again, ensure that the test succeeds without any warnings.
+%
+% Once you have done this, you have succeeded in creating a non-trivial Prolog
+% program that does something interesting.  Congratulations!
+%
+
+
+maze_solution(From,To,Path,Len) :-
+	connected(From,From),
+	connected(From,To),
+	connected(From,Path),
+	len(Len,[_|Path]).
+
```

Comment:
The solution should be easy to build from the steps taken above.  In particular, you just need to build a path from From to To, and then get its length.